### PR TITLE
FEATURE: Implement CodeMirror editor

### PR DESCRIPTION
**Pull request by kitsunet**

### CODE REVIEW

**Comment by kitsunet** [comment]
Apparently Resources/Public/ is gitignored, I guess that is intentional, do I force the codemirror css files?

**Comment by kitsunet** [comment]
Also adding the react-codemirror dependency to the editors package didn't work for me, had to add it to the neos-ui package

**Comment by kitsunet** [comment]
Mmm, what did I do? \o/

**Comment by skurfuerst** [comment]
@kitsunet you just have some linting errors: https://travis-ci.org/neos/neos-ui/jobs/203821320#L302

File: packages/neos-ui-editors/src/CodeMirror/CodeMirrorWrap/index.js [hunk 1]
```diff
@@ -0,0 +1,42 @@
+import React, {PureComponent, PropTypes} from 'react';
+import CodeMirror from 'react-codemirror';
+
+// TODO: Find way to dynamically load any mode?
+import XmlMode from 'codemirror/mode/xml/xml';
+import HtmlMixedMode from 'codemirror/mode/htmlmixed/htmlmixed';
+import JavascriptMode from 'codemirror/mode/javascript/javascript';
+import HandlebarsMode from 'codemirror/mode/handlebars/handlebars';
+import MarkdownMode from 'codemirror/mode/markdown/markdown';
+import YamlMode from 'codemirror/mode/yaml/yaml';
```
Comment (skurfuerst) [inline packages/neos-ui-editors/src/CodeMirror/CodeMirrorWrap/index.js]: in den Zeilen oben musst du ESLint ausschalten:

```
/* eslint-disable no-unused-vars */
...
/* eslint-ensable no-unused-vars */
```

File: packages/neos-ui-editors/src/CodeMirror/index.js [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    };
+
+    constructor(props) {
+        super(props);
+        this.handleOpenCodeEditor = this.handleOpenCodeEditor.bind(this);
+    }
+
+    handleOpenCodeEditor() {
+        this.props.renderSecondaryInspector('CODEMIRROR_EDIT', () =>
+            <CodeMirrorWrap onChange={this.props.commit} value={this.props.value} highlightingMode={this.props.highlightingMode} />
```
Comment (skurfuerst) [inline packages/neos-ui-editors/src/CodeMirror/index.js]: this.props.commit ist "direkt" nicht erlaubt per code-convention -> ich würde nochmal in dem Fall ne Zwischenfunktion "handleCodeChange()" machen, die dann this.props.commit() triggert.

Comment (skurfuerst) [inline packages/neos-ui-editors/src/CodeMirror/index.js]: zusätzlich das Leerzeichen vor `/>` entfenren

File: packages/neos-ui-editors/src/CodeMirror/index.js [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    }
+
+    render() {
+        const {label, identifier} = this.props;
+
+        return (
+            <div>
+                <Label htmlFor={identifier}>
+                    <Button onClick={this.handleOpenCodeEditor} style="brand">
+                        <Icon icon="pencil" padded="right" style="lighter" title="Edit" />
```
Comment (skurfuerst) [inline packages/neos-ui-editors/src/CodeMirror/index.js]: zusätzlich das Leerzeichen vor `/>` entfenren

File: packages/neos-ui-editors/src/CodeMirror/CodeMirrorWrap/index.js [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+        highlightingMode: PropTypes.string.isRequired,
+        value: PropTypes.string
+    };
+
+    constructor(props) {
+        super(props);
+        this.handleChange = this.handleChange.bind(this);
+    }
+
+    handleChange(newValue) {
```
Comment (skurfuerst) [inline packages/neos-ui-editors/src/CodeMirror/CodeMirrorWrap/index.js]: I usually write these "handle..." methods *after* render(), as this reads better for me...

**Comment by kitsunet** [comment]
I am not getting along with this linter... What is the problem now? @grebaldi

File: packages/neos-ui-editors/src/CodeMirror/index.js [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+        );
+    }
+
+    handleChange(newValue) {
+        this.props.commit(newValue);
+    }
+
+    handleOpenCodeEditor() {
+        this.props.renderSecondaryInspector('CODEMIRROR_EDIT', () =>
+            <CodeMirrorWrap onChange={this.handleChange} value={this.props.value}
```
Comment (grebaldi) [inline packages/neos-ui-editors/src/CodeMirror/index.js]: If multiple lines are needed for passing the props, it is considered best practice to do this with all of them, so `onChange` should already be placed on a new line here.

Also, those props need to have a 4-space indentation relative to the opening bracket, whereas the closing bracket needs to be aligned with the last prop (on a new line).

**Comment by kitsunet** [comment]
I think I got it, but currently the highlighting is broken again...

**Comment by skurfuerst** [comment]
nice :)

**Comment by dimaip** [comment]
Guys, why didn't we use some React version of the CodeMirror, and include it standalone via a script tag?
E.g. this one https://github.com/JedWatson/react-codemirror

**Comment by kitsunet** [comment]
@dimaip that is exactly the one I used.

**Comment by dimaip** [comment]
@kitsunet lol, sorry, I just took a quick glance and obviously was blind :-)

File: Resources/Private/Templates/Backend/Index.html [hunk 6]
```diff
@@ -8,6 +8,8 @@
         <link rel="stylesheet" href="{f:uri.resource(path: 'Styles/Host.css', package: compiledResourcePackage, localize: false)}">
         <link rel="stylesheet" href="{f:uri.resource(path: 'Styles/HostOnlyStyles.css', package: compiledResourcePackage, localize: false)}">
         <link rel="stylesheet" href="{f:uri.resource(path: 'Styles/ReactCrop.css', localize: false)}">
+        <link rel="stylesheet" href="{f:uri.resource(path: 'Styles/codemirror.css', localize: false)}">
```
Comment (dimaip) [inline Resources/Private/Templates/Backend/Index.html]: Still @kitsunet , why do we manually add its CSS here like this, and not via webpack's CSS loader?

**Pull request merged**